File: services/app/dockers/ruby/checker_example.rb
# frozen_string_literal: true

require 'json'
require 'stringio'
require 'test/unit'

extend Test::Unit::Assertions

original_stdout = $stdout
$stdout = StringIO.new

@execution_result = []

begin
  require './solution_example'

  success = true

  def assert_result(solution, expected, arguments, success)
    start = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    result = solution.call(*arguments)
    finish = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    assert_equal(expected, result)

    @execution_result <<
      JSON.dump(
        status: :success,
        result: result,
        output: $stdout.string,
        expected: expected,
        arguments: arguments,
        execution_time: finish - start
      )
    $stdout.reopen
    success
  rescue Test::Unit::AssertionFailedError
    @execution_result <<
      JSON.dump(
        status: "failure",
        result: result,
        output: $stdout.string,
        expected: expected,
        arguments: arguments,
        execution_time: finish - start
      )
    $stdout.reopen
    false
  end

  success = assert_result(method(:solution), 3, [1, 2], success)
  success = assert_result(method(:solution), 8, [5, 3], success)
  if success
    @execution_result <<
      JSON.dump(
        status: :ok,
        result: '__code-0__'
      )
  end
rescue Exception => e
  @execution_result <<
    JSON.dump( status: "error", result: e.message)
end

$stdout = original_stdout
puts @execution_result

File: services/app/dockers/python/checker_example.py
import sys
import json
import time
from io import StringIO

original_stdout = sys.stdout
sys.stdout = StringIO()

execution_result = []

def assert_result(solution, expected, arguments, success):
    try:
        start = time.clock()
        result = solution(arguments)
        finish = time.clock()
        assert result == expected
        execution_result.append(json.dumps({
            'status': 'success',
            'result': result,
            'output': sys.stdout.getvalue(),
            'expected': expected,
            'arguments': arguments,
            'execution_time': finish - start
        }))
        return success
    except AssertionError as exc:
        execution_result.append(json.dumps({
            'status': 'failure',
            'result': result,
            'output': sys.stdout.getvalue(),
            'expected': expected,
            'arguments': arguments,
            'execution_time': finish - start
        }))
        return False

try:
    from solution_example import solution
    success = True

    solution_lambda = lambda arguments: solution(*arguments)
    success = assert_result(solution_lambda, 3, [1, 2], success)
    success = assert_result(solution_lambda, 7, [5, 3], success)

    if success:
        execution_result.append(json.dumps({
            'status': 'ok',
            'result': '__code-0__',
        }))
except Exception as exc:
    execution_result.append(json.dumps({
        'status': 'error',
        'result': exc.args,
    }))

sys.stdout = original_stdout
print(execution_result)
